Port ec2_elb to boto3 - fixes #30229

s-hertel opened the pull request:
##### SUMMARY
Fixes #30229 by adding AWSRetry.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/ec2_elb.py

##### ANSIBLE VERSION
```
2.5.0
```

s-hertel committed:
Port ec2_elb to boto3

s-hertel committed:
Integration tests for ec2_elb.

s-hertel commented:
Need to add permissions to Shippable and update the policy in hacking/aws_config.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_elb.py:
```diff
@@ -353,25 +377,39 @@ def main():
+    try:
+        elb_man.lbs = elb_man._get_instance_lbs(client, ec2_elbs)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e)
```
Please also provide a message here.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_elb.py:
```diff
@@ -295,34 +319,37 @@ def _get_auto_scaling_group_lbs(self):
-        return ec2.get_only_instances(instance_ids=[self.instance_id])[0]
+            ec2 = boto3_conn(self.module, conn_type='client', resource='ec2', region=self.region, **self.aws_connect_params)
+        except botocore.exceptions.ProfileNotFound as e:
+            self.module.fail_json_aws(e)
```
Message on fail please :smiley:

ryansb commented on test/integration/targets/ec2_elb/tasks/main.yml:
```diff
@@ -0,0 +1,536 @@
+---
+- block:
```
I think `block` has a `environment:` attribute that you can use to set all the aws secret/key id/session token variables as their bash appropriate versions (AWS_SECRET_KEY etc) and not have to do so on every module call. 

For the tests checking that the module fails correctly when missing credentials, put them in a separate block.

ryansb commented on test/integration/targets/ec2_elb/tasks/main.yml:
```diff
@@ -0,0 +1,536 @@
+    - name: assert task failed
+      assert:
+        that:
+          - "result.failed"
```
The preferred way to check failure is to use the filter, so there's no error if that attribute is undefined. Just `result | failed`

s-hertel committed:
Add specific failure messages to the ec2_elb module exception handling.

Reorganize integration tests for ec2_elb, using the block's environment to set credentials instead of specifying them for every task.

s-hertel committed:
Fix indentation

ElijahLynn commented:
I have been testing this for some provisioning over the past week and it has saved the day in terms of preventing API throttled errors/messages from AWS!!! So, :+1: for some basic testing so far.

wimnat commented:
I thought the plan for this was to just leave this module 'as-is' and instead create elb_classic_lb as the boto3 version allowing us to fix return values etc to be more in line with elb_application_lb?

s-hertel commented:
@wimnat Ah....Sigh. Yes, that was the plan. Thanks for noticing this!

s-hertel commented:
@wimnat Are you working on elb_classic_lb for 2.5?

wimnat commented:
I'd like to but not started yet. However my head is in the space now - just pushed WIP PR for elb_network_lb.

ElijahLynn commented:
Here is the WIP PR referenced by @wimnat https://github.com/ansible/ansible/pull/33808

jchristi commented:
The above linked PR is for [elb_network_lb](https://github.com/wimnat/ansible/blob/3150abae56f304f6091ae7d1eb42f782de1da501/lib/ansible/modules/cloud/amazon/elb_network_lb.py), not [elb_classic_lb](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/elb_classic_lb.py) or [ec2_elb_lb]((https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ec2_elb_lb.py)). A quick look at the above two modules show they are both still using old boto. Is there anywhere boto3 conversion for a classic elb is being tracked? @ElijahLynn and I really want this!

ElijahLynn commented:
@s-hertel Do you know if this functionality is somehow in the recent 2.6 release? I see https://github.com/ansible/ansible/blame/devel/lib/ansible/modules/cloud/amazon/elb_application_lb.py#L28 uses boto3 but did not see any reference of exponential or backoff, but that doesn't mean for sure it isn't included.

jchristi commented:
@s-hertel I have been using elb_application_lb and have yet to see any API throttling there. I doubt I will be able to find the time to test classic elb, but I'll let you know if I do.

jamiecwilliams commented:
> I thought the plan for this was to just leave this module 'as-is' and instead create elb_classic_lb as the boto3 version allowing us to fix return values etc to be more in line with elb_application_lb?

> @wimnat Ah....Sigh. Yes, that was the plan. Thanks for noticing this!

@s-hertel, isn't this PR for porting the `ec2_elb` module? The `ec2_classic_lb` module should not be relevant, because it is for managing the lbs, where `ec2_elb` is for registering instances. As far as I can tell, there is no alternative to `ec2_elb` which is boto3 based, and it is still lacking AWSRetry support. @ryansb added `ec2_elb` to the list of "boto only" modules very recently: https://github.com/ansible/community/blob/master/group-aws/boto3.md

Perhaps @wimnat confused your effort with decisions over the `ec2_elb_lb` module, which has been superceded (if I understand correctly) by `ec2_classic_lb`, `ec2_network_lb`, and `ec2_appliance_lb`?
Issue https://github.com/ansible/ansible/issues/30494 regarding renaming or replacing existing lb-related modules, suggested `ec2_instance` as a new boto3 module. However, `ec2_instance` is also in the list of "boto only" modules. The current documentation does not indicate any retry capability in this new module either.

I'm having major difficulty due to throttling with the `ec2_elb` module. Is there any reason `elb_instance` would reduce throttling problems? What is the difference between `ec2_elb` and `ec2_instance` if they both use boto? Is there still a plan to update either to boto3?

Since issue https://github.com/ansible/ansible/issues/30229 references this PR and is still open, and this is still a current problem with this module, I request that this PR be re-opened or similar changes be made to the `ec2_instance` module.
